proposals/0000-import-shadowing.rst

```diff
@@ -0,0 +1,430 @@
+.. sectnum::
+.. contents::
+
+Local binders are allowed to shadow names defined in outer
+scopes. Here, we propose that top-level binders should also be allowed
+to shadow imported names, basically treating imported modules as
+living in a scope that is "outside" the top-level one.
+
+
+Motivation
+----------
+Currently, there is no way in Haskell to shadow an imported name in
+the top-level scope. The only workaround is to avoid importing said
+name at all. For example, if we want to use some ``Control.Exception``
+functions while also defining our own ``catch``, we have to import the
+former with an explicit hiding of ``catch``:
+
+::
+
+ module Example1 where
+
+ import Control.Exception hiding (catch)
+
+ catch :: Fish -> IO Food
+ catch fish = ...
+
+ goFishing :: IO Food
+ goFishing = bracket goToCabin goHome $ catch salmon
+
+Without hiding ``catch``, this results in a name resolution conflict
+in the definition of ``goFishing``:
+
+::
+
+ Ambiguous occurrence `catch'
+ It could refer to
+    either `Control.Exception.catch',
+           imported from `Control.Exception' at Example1.hs:1:1-24
+           (and originally defined in `GHC.IO')
+        or `Example1.catch', defined at Example1.hs:7:1
+
+Our proposal is an alternative name resolution policy exposed as a
+language extension tentatively named ``ImportShadowing``. It allows
+definitions from the current module's top-level scope to shadow names
+from imported modules. So in this example, the occurrence of ``catch``
+in the definition of ``goFishing`` resolves to the definition in
+``Example1`` even when ``catch`` is not explicitly hidden from the
+import list of ``Control.Exception``.
+
+We believe this is an improvement to the status quo because it is a
+natural extension of the idea behind the shadowing policy of local
+binders: that shadowing expresses the intention by the user to only
+care about the names that are defined "more nearby". A cursory search
+on `Stack Overflow <https://stackoverflow.com/>`_ finds lots of
+Haskell users who implicitly expected imports to be shadowed by
+top-level definitions:
+
+* https://stackoverflow.com/q/73788349/477476
+* https://stackoverflow.com/q/7761238/477476
+* https://stackoverflow.com/q/56047335/477476
+* https://stackoverflow.com/q/40964909/477476
+* https://stackoverflow.com/q/40314142/477476
+* https://stackoverflow.com/q/67246392/477476
+
+A second-order effect of the proposed extension is that it can lead to
+preemptive forward compatibility. Adding a new export to ``Prelude``
+can lead to breakage just by virtue of existing code defining and
+using top-level definitions with the same name. With
+``ImportShadowing``, the existing intra-module references keep their
+meaning and there is no migration needed to accomodate the new
+``Prelude`` names.
+
+Proposed Change Specification
+-----------------------------
+
+When ``ImportShadowing`` is enabled, the following changes take place:
```

Comment:
You say so in your motivation section, but for clarity, please state clearly in this section that `ImportShadowing` is a new extension that this proposal creates.